### ISSUE TITLE
First graphql Channel test via assert_reply times out

### ISSUE THREAD

**Issue by dustinfarris** [absinthe-graphql/absinthe#289]
Reposting from https://github.com/benwilson512/phoenix_chat_example/pull/2

Also, see: https://github.com/phoenixframework/phoenix/issues/2023

Discussions with @bruce and @benwilson512 and @josevalim indicate that this is due to the sheer size of absinthe.  When it is loaded for the first time, the first test times out.

----

### Summary

The `assert_reply` channel test macro times out on the _first_ test that invokes `Absinthe.run` and returns a result.

### Details

This PR contains models for `Chat.Message` and `Chat.User`, a new schema field `message`, and a "graphql" channel that takes a standard GraphQL query, runs it through Absinthe, and returns the result.

There is a test that verifies that this works.  The test creates a Message, sends a message query to the graphql channel, and listens for a reply using the `assert_reply` macro.  `assert_reply`, by default, times out after 100ms.

There are actually two tests that are nearly identical.  The only difference is the test on line 44 uses `assert_reply`'s default timeout.  The test on line 28 increases the timeout to 300ms.

I am using Elixir 1.4.0 and Erlang 19.2

#### To reproduce

Clone this repo:

```
git clone git@github.com:dustinfarris/phoenix_chat_example
cd phoenix_chat_example && git checkout channel-timeout
mix do deps.get, compile
```

Run the default timeout test:

```
mix test test/message_query_test.exs:44
```

The test fails:

```
  1) test message query (Chat.MessageQueryTest)
     test/message_query_test.exs:44
     No message matching %Phoenix.Socket.Reply{ref: ^ref, status: :ok, payload: %{data: %{"message" => %{"body" => "Test message", "author" => %{"name" => "Dustin Farris"}}}}} after 100ms.
     The following variables were pinned:
       ref = #Reference<0.0.2.17387>
     The process mailbox is empty.
     stacktrace:
       test/message_query_test.exs:50: (test)
```

That "The process mailbox is empty." line is a result of the assertion timing out before a message was received.

Now, run the test where the timeout is increased to 300ms

```
mix test test/message_query_test.exs:28
```

The test passes:

```
.

Finished in 0.2 seconds
```

Now—and this is interesting—run both tests:

```
mix test test/message_query_test.exs
```

Both tests pass:

```
..

Finished in 0.2 seconds
```

----

I have observed that only the first test that runs will fail with this timeout error, if it is left at the default 100ms.  From that point on everything passes.

**Comment by bruce** [comment]
"Absinthe is big" doesn't seem to have an actionable... solution.

**Comment by bruce** [comment]
There's also not any "lazy compilation" in play, so this seems very strange to me. I've certainly never seen this behavior myself. @benwilson512?

**Comment by benwilson512** [comment]
@bruce The basic issue is just that mix loads code lazily for tests, and Absinthe has a lot of code. This code isn't loaded until the first query is run, and so that first query takes exceptional time.

My proposal is that we supply an `Absinthe.Test.prime(MyApp.Schema)` function that you place in your `test_helpers.exs` file that just runs the introspection query against the user's schema. This should load the vast majority of the Absinthe code prior to any actual test cases.

**Comment by dustinfarris** [comment]
@bruce this is not an issue for people doing things over HTTP (majority, I imagine); but, since I am doing everything over Channels, I am using the Phoenix [`assert_reply`](https://hexdocs.pm/phoenix/Phoenix.ChannelTest.html#assert_reply/4) helper to check the graphql responses—which operates on a timeout.  The first test spends the default allotted time (100ms) loading Absinthe.

Apologies for the trolling title.  I have not gotten much sleep.

**Comment by bruce** [comment]
@dustinfarris Not a problem at all -- I was unaware of the details here (honestly, I should have researched further before responding). I'm glad @benwilson512 has a solution in mind.

**Comment by benwilson512** [comment]
@dustinfarris Are you able to use the 1.3 branch? We have added https://github.com/absinthe-graphql/absinthe/blob/v1.3/lib/absinthe/test.ex#L21

**Comment by dustinfarris** [comment]
@benwilson512 did not have any effect :-(

I tried adding `prime` to test_helpers, channel_case, and the test itself.

**Comment by dustinfarris** [comment]
Some times:

Added to graphql channel:

```
   14   def handle_in("query", %{"query" => query, "variables" => variables}, socket) do
   15     {time, _} = :timer.tc(fn ->
   16       Absinthe.run(Schema, variables: variables, context: build_context(socket))
   17     end)
   18
   19     IO.puts "+++++++++++++++++++++++++++++++++++++++"
   20     IO.puts IO.inspect(time)
   21     IO.puts "+++++++++++++++++++++++++++++++++++++++"
```

Added to test_helper.exs

```
Absinthe.Test.prime(Schema)
```

Output from running a handful of tests:

```
+++++++++++++++++++++++++++++++++++++++
17361
17361
+++++++++++++++++++++++++++++++++++++++
.+++++++++++++++++++++++++++++++++++++++
139
139
+++++++++++++++++++++++++++++++++++++++
.+++++++++++++++++++++++++++++++++++++++
153
153
+++++++++++++++++++++++++++++++++++++++
.+++++++++++++++++++++++++++++++++++++++
151
151
+++++++++++++++++++++++++++++++++++++++
```

So, that first test would have failed if I didn't manually increase the timeout.

**Comment by josevalim** [comment]
Maybe the schema is causing Ecto to be loaded and so on? Maybe it is better to trigger a fake request over the channel?

**Comment by dustinfarris** [comment]
@josevalim yeah that works

threw this at the top of the test module

```elixir
  @introspection_query """
    {
      __schema {
        types {
          name
        }
      }
    }
  """

  setup context do
    context = graphql_socket(context)

    ref = push context.socket, "query", %{
      query: @introspection_query,
      variables: %{}
    }

    assert_reply ref, :ok, %{data: %{"__schema" => _}}, 900

    context
  end
```

and everything passes.  @benwilson512 so it seems that the channel process itself needs to be "primed"?   yet when I put `prime` in the join function of the channel the tests still fail.  so maybe `prime` isn't doing what we think it is?

**Comment by benwilson512** [comment]
I had an obvious bug in the prime function and it wasn't actually running stuff sorry. Can you try the `v1.3` branch?

**Comment by dustinfarris** [comment]
It works!

Thanks, Ben!

tl;dr - you can add this to your test_helper.exs

```elixir
Absinthe.Test.prime(MyApp.Schema)
```
